main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+      }
+    }
+
+  }
+
+  /**
+    * Returns a list of non-trivial unification equations computed from the given list `l`.
+    *
+    * Throws a [[ConflictException]] if an unsolvable equation is encountered.
+    *
+    * A trivial equation is one of:
+    * -     true ~ true
+    * -    false ~ false
+    * -        c ~ c        (same constant)
+    * -        x ~ x        (same variable)
+    *
+    *
+    * An unsolvable (conflicted) equation is one of:
+    *
+    * -      true ~ false   (and mirrored)
+    * -       c_i ~ c_j     (different constants)
+    * -        c ~ true     (and mirrored)
+    * -        c ~ false    (and mirrored)
+    */
+  private def simplify(l: List[Equation]): List[Equation] = l match {
+    case Nil => Nil
+    case Equation(t1, t2, loc) :: es => (t1, t2) match {
+      // Trivial equations: skip them.
+      case (Term.True, Term.True) => simplify(es)
+      case (Term.False, Term.False) => simplify(es)
+      case (Term.Cst(c1), Term.Cst(c2)) if c1 == c2 => simplify(es)
+      case (Term.Var(x1), Term.Var(x2)) if x1 == x2 => simplify(es)
+
+      // Unsolvable (conflicted) equations: raise an exception.
+      case (Term.True, Term.False) => throw ConflictException(t1, t2, loc)
+      case (Term.False, Term.True) => throw ConflictException(t1, t2, loc)
+      case (Term.Cst(c1), Term.Cst(c2)) if c1 != c2 => throw ConflictException(t1, t2, loc)
+      // Note: A constraint with two different variables is of course solvable!
+      case (Term.Cst(_), Term.True) => throw ConflictException(t1, t2, loc)
+      case (Term.Cst(_), Term.False) => throw ConflictException(t1, t2, loc)
+      case (Term.True, Term.Cst(_)) => throw ConflictException(t1, t2, loc)
+      case (Term.False, Term.Cst(_)) => throw ConflictException(t1, t2, loc)
+
+      // Non-trivial and non-conflicted equation: keep it.
+      case _ => Equation(t1, t2, loc) :: simplify(es)
+    }
+  }
+
+  /**
+    * Propagates constants and truth values through the equation system.
+    *
+    * The implementation saturates the system, i.e. it computes a fixpoint.
+    *
+    * The implementation uses three rewrite rules:
+    *
+    * - `x ~ true` becomes `[x -> true]`.
+    * - `x ~ c` becomes `[x -> c]`.
+    * - `x /\ y /\ ... = true` becomes `[x -> true, y -> true, ...]`.
+    *
+    * For example, if the equation system is:
+    *
+    * {{{
+    *     c1794221043 ~ (x55062 ∧ x55050 ∧ x55046 ∧ x55060 ∧ x55066 ∧ x55040 ∧ x55075 ∧ x55042 ∧ x55058 ∧ x55078)
+    *     x55078 ~ x112431
+    *     c1794221043 ~ x55040
+    *     x55042 ~ x112433
+    *     x55044 ~ x112437
+    *     x55046 ~ (x112435 ∧ x55044)
+    *     x55048 ~ true
+    *     x55050 ~ (x112439 ∧ x55048)
+    *     x55052 ~ x112443
+    *     x55055 ~ true
+    *     x55058 ~ (x55052 ∧ x112441 ∧ x55055)
+    *     x55060 ~ x112446
+    *     x55062 ~ x112448
+    *     x55066 ~ true
+    *     x55075 ~ x112453
+    * }}}
+    *
+    * then after unit propagation it is:
```

Comment:
This is the first time unit propagation is called that, maybe call it unit propagation everywhere or write constant propagation here.